server/src/main/java/org/elasticsearch/cluster/coordination/CoordinationState.java

```diff
@@ -0,0 +1,485 @@
+            logger.debug("handleJoin: ignored join due to term mismatch (expected: [{}], actual: [{}])",
+                getCurrentTerm(), join.getTerm());
+            throw new CoordinationStateRejectedException(
+                "incoming term " + join.getTerm() + " does not match current term " + getCurrentTerm());
+        }
+
+        if (startedJoinSinceLastReboot == false) {
+            logger.debug("handleJoin: ignored join as term was not incremented yet after reboot");
+            throw new CoordinationStateRejectedException("ignored join as term has not been incremented yet after reboot");
+        }
+
+        final long lastAcceptedTerm = getLastAcceptedTerm();
+        if (join.getLastAcceptedTerm() > lastAcceptedTerm) {
+            logger.debug("handleJoin: ignored join as joiner has a better last accepted term (expected: <=[{}], actual: [{}])",
+                lastAcceptedTerm, join.getLastAcceptedTerm());
+            throw new CoordinationStateRejectedException("incoming last accepted term " + join.getLastAcceptedTerm() +
+                " of join higher than current last accepted term " + lastAcceptedTerm);
+        }
+
+        if (join.getLastAcceptedTerm() == lastAcceptedTerm && join.getLastAcceptedVersion() > getLastAcceptedVersion()) {
+            logger.debug("handleJoin: ignored join as joiner has a better last accepted version (expected: <=[{}], actual: [{}])",
+                getLastAcceptedVersion(), join.getLastAcceptedVersion());
+            throw new CoordinationStateRejectedException("incoming last accepted version " + join.getLastAcceptedVersion() +
+                " of join higher than current last accepted version " + getLastAcceptedVersion());
+        }
+
+        if (getLastAcceptedVersion() == 0) {
+            // We do not check for an election won on setting the initial configuration, so it would be possible to end up in a state where
+            // we have enough join votes to have won the election immediately on setting the initial configuration. It'd be quite
+            // complicated to restore all the appropriate invariants when setting the initial configuration (it's not just electionWon)
+            // so instead we just reject join votes received prior to receiving the initial configuration.
+            logger.debug("handleJoin: ignored join because initial configuration not set");
+            throw new CoordinationStateRejectedException("initial configuration not set");
+        }
+
+        boolean added = joinVotes.addVote(join.getSourceNode());
+        boolean prevElectionWon = electionWon;
+        electionWon = isElectionQuorum(joinVotes);
+        logger.debug("handleJoin: added join {} from [{}] for election, electionWon={} lastAcceptedTerm={} lastAcceptedVersion={}", join,
+            join.getSourceNode(), electionWon, lastAcceptedTerm, getLastAcceptedVersion());
+
+        if (electionWon && prevElectionWon == false) {
```

Comment:
Tests all pass with `if (electionWon) {`

Reply:
fixed in 51192562026 (which also found another small copy-paste error)